plugin/tls/setup.go

```diff
@@ -0,0 +1,209 @@
+package tls
+
+import (
+	"context"
+	ctls "crypto/tls"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/plugin/pkg/tls"
+
+	"github.com/caddyserver/certmagic"
+)
+
+func init() { plugin.Register("tls", setup) }
+
+func setup(c *caddy.Controller) error {
+	err := parseTLS(c)
+	if err != nil {
+		return plugin.Error("tls", err)
+	}
+	return nil
+}
+
+var (
+	log            = clog.NewWithPlugin("tls")
+	r              = renewCert{quit: make(chan bool), renew: make(chan bool)}
+	once, shutOnce sync.Once
+)
+
+const (
+	defaultCA            = "https://acme-v02.api.letsencrypt.org/directory"
+	defaultEmail         = "doesnotexist@test.com"
+	defaultCheckInterval = 15
+	defaultPort          = 53
+	defaultCertPath      = "./local/share/certmagic"
```

Comment:
Wondering if this is a generic path we want to use? What about non-Linux platforms? Can we just save the cert file within the same address as coredns binary or Corefile?

Reply:
This is what `certmagic` uses by default. Saving it in the same location as the coredns binary would be possible. 
Since there are a couple more files than just one cert, how about creating a directory `tls` - then we would end up with a directory structure like this:

```
.
├── Corefile
├── coredns
└── tls
    ├── acme
    │   ├── acme-staging-v02.api.letsencrypt.org-directory
    │   │   └── users
    │   │       └── doesnotexist@test.com
    │   │           ├── doesnotexist.json
    │   │           └── doesnotexist.key
    │   └── acme-v02.api.letsencrypt.org-directory
    │       ├── challenge_tokens
    │       └── users
    │           └── test@test.com
    │               ├── test.json
    │               └── test.key
    ├── certificates
    │   └── acme-staging-v02.api.letsencrypt.org-directory
    │       └── ns1.subdomain.mariuskimmina.com
    │           ├── ns1.subdomain.mariuskimmina.com.crt
    │           ├── ns1.subdomain.mariuskimmina.com.json
    │           └── ns1.subdomain.mariuskimmina.com.key
    ├── locks
    └── ocsp
        └── ns1.subdomain.mariuskimmina.com-a0e02e21
```

All directorys and filenames inside `tls` are then created by `certmagic`

Reply:
@mariuskimmina  sounds good, probably '.tls' to avoid possible collision?